README.md
# Format Preserving Encryption in Python

An implementation of the NIST-approved FF1 algorithm in Python.

This implementation conforms (as best as possible) to
[Draft SP 800-38G Rev. 1][800-38g1]. The implementation passes all tests
specified by NIST in their Cryptographic Standards and Guidelines
[examples for FF1][ff1-examples]


## Documentation

See the [Python API docs](https://dev.ubiqsecurity.com/docs/api).

## Installation

#### Using the package manager:
You may want to make sure you are running the latest version of pip3 by
first executing
```sh
$ pip3 install --upgrade pip
```

You don't need this source code unless you want to modify the package. If you just want to use the package, install from PyPi using pip3, a package manager for Python3.

```sh
$ pip3 install --upgrade ubiq_security_fpe
```


#### Installing from source:
From within the cloned git repository directory, Install from source with:


```
$ cd ubiq-fpe-python
$ pip3 install -r requirements.txt
$ python3 setup.py install
```
You may need to run the python3 commands above using sudo.

# Testing

To run the tests:

```sh
$ python3 -m unittest  ubiq_security_fpe/*test.py -v
```


### Requirements

-   Python 3.5+


### About alphabets and the radix parameter

The interfaces operate on strings, and the radix parameter determines which
characters are valid within those strings, i.e. the alphabet. For example, if
your radix is 10, then the alphabet for your plain text consists of the
characters in the string "0123456789". If your radix is 16, then the
alphabet is the characters in the string "0123456789abcdef".

More concretely, if you want to encrypt, say, a 16 digit number grouped into
4 groups of 4 using a `-` as a delimiter as in `0123-4567-8901-2345`, then you
would need a radix of at least 11, and you would need to translate the `-`
character to an `a` (as that is the value that follows `9`) prior to the
encryption. Conversely, you would need to translate an `a` to a `-` after
decryption.

This mapping of user inputs to alphabets defined by the radix is not performed
by the library and must be done prior to calling the encrypt and after calling
the decrypt functions.

A radix of up to 36 is supported, and the alphabet for a radix of 36 is
"0123456789abcdefghijklmnopqrstuvwxyz".

### Tweaks

Tweaks are very much like Initialization Vectors (IVs) in "traditional"
encryption algorithms. For FF1, the minimun and maximum allowed lengths of
the tweak may be specified by the user, and any tweak length between those
values may be used.

### Plain/ciphertext input lengths

For FF1, the minimum length is determined by the inequality:
- radix<sup>minlen</sup> >= 1000000

or:
- minlen >= 6 / log<sub>10</sub> radix

Thus, the minimum length is determined by the radix and is automatically
calculated from it.

For FF1, the maximum input length is
- 2<sup>32</sup>

## Examples

The [unit test code](ff1_test.py) provides the best
and simplest example of how to use the interfaces.

### FF1
```python
    /*
     * @key is a byte array whose length must be 16, 24, or 32
     * @twk is an optional byte array that can be used for the FF1 context
     * @twk_min_len is a constraint that can be applied for the tweak during
     *      the encrypt or decrypt calls
     * @twk_max_len is a constraint that can be applied for the tweak during
     *      the encrypt or decrypt calls
     * @radix is the radix for the text string
     * @alpha is the character set to use when perfoming the FPE encryption
     */

    ctx = ff1.Context(bytes(key), bytes(twk), twk_min_len, twk_max_len, radix, alpha)

    /*
     * @pt is the plain text string where the characters correspond to the 
     *     alpha character set and the radix value
     * @tweak is an optional byte array that will override the twk value
     *     supplied in the Context function.  If supplied it must be constrained to
     *     the twk_min_len and twk_max_len values
     */


    ct = ctx.Encrypt(pt, tweak)
    out = ctx.Decrypt(ct, tweak)
```



[dashboard]:https://dashboard.ubiqsecurity.com/
[credentials]:https://dev.ubiqsecurity.com/docs/how-to-create-api-keys
[800-38g1]:https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38Gr1-draft.pdf
[ff1-examples]:https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/FF1samples.pdf



CHANGELOG.md
# Changelog

## 2.0.1 - 2023-11-16
* Syntax change to support earlier versions of python

## 2.0.0 - 2023-04-24
* Standardize versioning

## 1.0.0 - 2023-01-26
* Initial Version


test.py
import unittest

import ff1
import ffx

class TestFFX(unittest.TestCase):
    def test_context(self):
        self.assertIsNotNone(
            ffx.Context(bytes([0]*32), bytes([0]*7), 2**32, 0, 7, 10))

class TestFF1(unittest.TestCase):
    def test_context(self):
        self.assertIsNotNone(
            ff1.Context(bytes([0]*16), bytes([0]*7), 0, 7, 10))

    def cipherTest(self, key, twk, pt, ct, radix):
        ctx = ff1.Context(bytes(key), bytes(twk), 0, len(twk), radix)
        self.assertIsNotNone(ctx)
        res = ctx.Encrypt(pt, None)
        self.assertEqual(res, ct)
        res = ctx.Decrypt(ct, None)
        self.assertEqual(res, pt)

    def test_nist1(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [],
            '0123456789', '2433477484', 10)

    def test_nist2(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [0x39, 0x38, 0x37, 0x36, 0x35, 0x34, 0x33, 0x32, 0x31, 0x30],
            '0123456789', '6124200773', 10)

    def test_nist3_prime(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [0x37, 0x37, 0x37, 0x37, 0x70, 0x71, 0x72, 0x73, 0x37, 0x37, 0x37],
            '0123456789abcdefghi', 'a9tv40mll9kdu509eum', 36)

    def test_nist3_prime(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [0x37, 0x37, 0x37, 0x37, 0x70, 0x71, 0x72, 0x73, 0x37, 0x37, 0x37],
            '0123456789abcdefghijklmnopqrstuvwxyz012345',
            'l4zpm8zauhw2zkke9o0i7oi17iojfqog7dmjze3her',
            36)

if __name__ == '__main__':
    unittest.main()


ff1_test.py
#!/usr/bin/env python3

import unittest

import importlib
ff1 = importlib.import_module('ubiq-fpe-python.ff1')
ffx = importlib.import_module('ubiq-fpe-python.ffx')

class TestFF1(unittest.TestCase):
    def test_context(self):
        self.assertIsNotNone(
            ff1.Context(bytes([0]*16), bytes([0]*7), 0, 7, 10))

    def cipherTest(self, key, twk, pt, ct, radix, alpha = ffx.DEFAULT_ALPHABET):
        ctx = ff1.Context(bytes(key), bytes(twk), 0, len(twk), radix, alpha)
        self.assertIsNotNone(ctx)
        res = ctx.Encrypt(pt)
        self.assertEqual(res, ct)
        res = ctx.Decrypt(ct)
        self.assertEqual(res, pt)

    def test_nist1(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [],
            '0123456789', '2433477484', 10)

    def test_nist2(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [0x39, 0x38, 0x37, 0x36, 0x35, 0x34, 0x33, 0x32, 0x31, 0x30],
            '0123456789', '6124200773', 10)

    def test_nist2_custom(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [0x39, 0x38, 0x37, 0x36, 0x35, 0x34, 0x33, 0x32, 0x31, 0x30],
            '2345678901', '8346422995', 10, '2345678901')

    def test_nist3(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [0x37, 0x37, 0x37, 0x37, 0x70, 0x71, 0x72, 0x73, 0x37, 0x37, 0x37],
            '0123456789abcdefghi', 'a9tv40mll9kdu509eum', 36)

    def test_nist3_prime(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c],
            [0x37, 0x37, 0x37, 0x37, 0x70, 0x71, 0x72, 0x73, 0x37, 0x37, 0x37],
            '0123456789abcdefghijklmnopqrstuvwxyz012345',
            'l4zpm8zauhw2zkke9o0i7oi17iojfqog7dmjze3her',
            36)

    def test_nist4(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c,
             0xef, 0x43, 0x59, 0xd8, 0xd5, 0x80, 0xaa, 0x4f],
            [],
            '0123456789', '2830668132', 10)

    def test_nist5(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c,
             0xef, 0x43, 0x59, 0xd8, 0xd5, 0x80, 0xaa, 0x4f],
            [0x39, 0x38, 0x37, 0x36, 0x35, 0x34, 0x33, 0x32, 0x31, 0x30],
            '0123456789', '2496655549', 10)

    def test_nist6(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c,
             0xef, 0x43, 0x59, 0xd8, 0xd5, 0x80, 0xaa, 0x4f],
            [0x37, 0x37, 0x37, 0x37, 0x70, 0x71, 0x72, 0x73, 0x37, 0x37, 0x37],
            '0123456789abcdefghi', 'xbj3kv35jrawxv32ysr', 36)

    def test_nist7(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c,
             0xef, 0x43, 0x59, 0xd8, 0xd5, 0x80, 0xaa, 0x4f,
             0x7f, 0x03, 0x6d, 0x6f, 0x04, 0xfc, 0x6a, 0x94],
            [],
            '0123456789', '6657667009', 10)

    def test_nist8(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c,
             0xef, 0x43, 0x59, 0xd8, 0xd5, 0x80, 0xaa, 0x4f,
             0x7f, 0x03, 0x6d, 0x6f, 0x04, 0xfc, 0x6a, 0x94],
            [0x39, 0x38, 0x37, 0x36, 0x35, 0x34, 0x33, 0x32, 0x31, 0x30],
            '0123456789', '1001623463', 10)

    def test_nist9(self):
        self.cipherTest(
            [0x2b, 0x7e, 0x15, 0x16, 0x28, 0xae, 0xd2, 0xa6,
             0xab, 0xf7, 0x15, 0x88, 0x09, 0xcf, 0x4f, 0x3c,
             0xef, 0x43, 0x59, 0xd8, 0xd5, 0x80, 0xaa, 0x4f,
             0x7f, 0x03, 0x6d, 0x6f, 0x04, 0xfc, 0x6a, 0x94],
            [0x37, 0x37, 0x37, 0x37, 0x70, 0x71, 0x72, 0x73, 0x37, 0x37, 0x37],
            '0123456789abcdefghi', 'xs8a0azh2avyalyzuwd', 36)


ffx_test.py
#!/usr/bin/env python3

import unittest

import importlib
ffx = importlib.import_module('ubiq-fpe-python.ffx')

class TestFFX(unittest.TestCase):
    def test_context(self):
        self.assertIsNotNone(
            ffx.Context(bytes([0]*32), bytes([0]*7),
                        2**32,
                        0, 7,
                        10, ffx.DEFAULT_ALPHABET))


ff1.py
#!/usr/bin/env python3

import math

from . import ffx

class Context:
    def __init__(self,
                 key, twk,
                 mintwklen, maxtwklen,
                 radix, alpha = ffx.DEFAULT_ALPHABET):
        self.ffx = ffx.Context(key, twk,
                               2**32,
                               mintwklen, maxtwklen,
                               radix, alpha)

    def cipher(self, X, T, ENC):
        BLKSZ = self.ffx.BLKSZ

        n = len(X)
        u = int(n / 2)
        v = n - u

        b = int((math.ceil(math.log2(self.ffx.radix) * v) + 7) / 8)
        d = 4 * int((b + 3) / 4) + 4

        R = [0] * int((d + (BLKSZ - 1)) / BLKSZ) * BLKSZ

        if T == None:
            T = self.ffx.twk
        if T == None:
            T = bytes([])

        if (n < self.ffx.mintxtlen or
            n > self.ffx.maxtxtlen or
            len(T) < self.ffx.mintwklen or
            (self.ffx.maxtwklen > 0 and
             len(T) > self.ffx.maxtwklen)):
            raise RuntimeError('Input or tweak length error')

        PQ = [0] * (BLKSZ + int((len(T) + b + 1 + 15) / BLKSZ) * BLKSZ)

        # initialize the P portion of PQ
        PQ[:8] = [1, 2, 1,
                  self.ffx.radix >> 16 & 0xff,
                  self.ffx.radix >> 8 & 0xff,
                  self.ffx.radix & 0xff,
                  10, u & 0xff]
        PQ[8:12] = n.to_bytes(4, byteorder='big')
        PQ[12:16] = len(T).to_bytes(4, byteorder='big')

        # initialize the constant portion of Q
        PQ[BLKSZ:BLKSZ + len(T)] = T

        nA = ffx.StringToNumber(self.ffx.radix, self.ffx.alpha, X[:u])
        nB = ffx.StringToNumber(self.ffx.radix, self.ffx.alpha, X[u:])
        if not ENC:
            nA, nB = nB, nA

        mU = self.ffx.radix ** u
        mV = mU
        if u != v:
            mV *= self.ffx.radix

        for i in range(10):
            if ENC:
                PQ[-b - 1] = i
            else:
                PQ[-b - 1] = 9 - i

            PQ[-b:] = nB.to_bytes(b, byteorder='big')

            R[:BLKSZ] = list(self.ffx.PRF(bytes(PQ)))

            for j in range(int(len(R) / BLKSZ) - 1):
                w = int.from_bytes(R[12:BLKSZ], byteorder='big')

                R[12:BLKSZ] = (w ^ (j + 1)).to_bytes(4, byteorder='big')
                R[BLKSZ * (j + 1):BLKSZ * (j + 2)] = list(
                    self.ffx.Ciph(bytes(R[:BLKSZ])))
                R[12:BLKSZ] = w.to_bytes(4, byteorder='big')

            y = int.from_bytes(R[:d], byteorder='big')
            if ENC:
                y = nA + y
            else:
                y = nA - y

            nA, nB = nB, nA

            if int(ENC) == i % 2:
                nB = y % mV
            else:
                nB = y % mU

        if not ENC:
            nA, nB = nB, nA

        return (ffx.NumberToString(self.ffx.radix, self.ffx.alpha, nA, u) +
                ffx.NumberToString(self.ffx.radix, self.ffx.alpha, nB, v))

    def Encrypt(self, pt, twk = None):
        return self.cipher(pt, twk, True)

    def Decrypt(self, ct, twk = None):
        return self.cipher(ct, twk, False)


ubiq_security_fpe/ffx.py
#!/usr/bin/env python3

import math
import sys
import typing

from M2Crypto import EVP

DEFAULT_ALPHABET: typing.Final[str] = '0123456789abcdefghijklmnopqrstuvwxyz'

class Context:
    def __init__(self,
                 key, twk,
                 maxtxtlen, mintwklen, maxtwklen,
                 radix, alpha):
        self.BLKSZ = (int)(16)
        self.key = key

        self.alg = None
        match len(key):
            case 16:
                self.alg = 'aes_128_cbc'
            case 24:
                self.alg = 'aes_192_cbc'
            case 32:
                self.alg = 'aes_256_cbc'
        if self.alg == None:
            raise RuntimeError('Key length invalid.')

        if radix < 2 or radix > len(alpha):
            raise RuntimeError('Unsupported radix or incompatible alphabet')

        self.alpha = alpha

        #
        # for both ff1 and ff3-1: radix**minlen >= 1000000
        #
        # therefore:
        #   minlen = ceil(log_radix(1000000))
        #          = ceil(log_10(1000000) / log_10(radix))
        #          = ceil(6 / log_10(radix))
        #
        mintxtlen = math.ceil(6 / math.log10(radix));
        if mintxtlen < 2 or mintxtlen > maxtxtlen:
            raise RuntimeError('Invalid text length bounds')

        if (mintwklen > maxtwklen or
            len(twk) < mintwklen or
            (maxtwklen > 0 and len(twk) > maxtwklen)):
            raise RuntimeError('Invalid tweak length or bounds')

        self.radix = radix

        self.mintxtlen = mintxtlen
        self.maxtxtlen = maxtxtlen
        self.mintwklen = mintwklen
        self.maxtwklen = maxtwklen

        self.twk = twk

    def PRF(self, buf):
        BLKSZ = self.BLKSZ

        if len(buf) % BLKSZ != 0:
            raise RuntimeError(
                'Plaintext length must be a multiple of ' +
                str(BLKSZ))

        cipher = EVP.Cipher(alg=self.alg, key=self.key, iv=bytearray(16), op=1)
        dst = cipher.update(buf)  

        return dst[-BLKSZ:]

    def Ciph(self, buf):
        return self.PRF(buf[0:self.BLKSZ])

def StringToNumber(radix, alpha, s):
    p = 1
    n = 0
    for i in range(len(s)):
        x = alpha.index(s[len(s) - i - 1])
        n += x * p
        p *= radix
    return n

def NumberToString(radix, alpha, n, l = 1):
    s = ''
    while n:
        s = alpha[int(n % radix)] + s
        n //= radix
    while len(s) < l:
        s = alpha[0] + s
    return s
